src/ansi-c/c_typecheck_expr.cpp

```diff
@@ -2801,6 +2803,86 @@ exprt c_typecheck_baset::do_special_functions(
     overflow.type() = bool_typet{};
     return overflow;
   }
+  else if(
+    identifier == "__builtin_add_overflow" ||
+    identifier == "__builtin_sub_overflow" ||
+    identifier == "__builtin_mul_overflow")
+  {
+    if(expr.arguments().size() != 3)
+    {
+      std::ostringstream error_message;
+      error_message << expr.source_location().as_string() << ": " << identifier
+                    << " takes exactly 3 arguments, but "
+                    << expr.arguments().size() << " were provided";
+      throw invalid_source_file_exceptiont{error_message.str()};
+    }
+
+    for(std::size_t arg_index = 0; arg_index < 3; ++arg_index)
+    {
+      auto const &argument = expr.arguments()[arg_index];
+      bool const arg_has_correct_type =
+        (arg_index < 2) ? (argument.type().id() == ID_signedbv ||
+                           argument.type().id() == ID_unsignedbv)
+                        : (argument.type().id() == ID_pointer &&
+                           (argument.type().subtype().id() == ID_signedbv ||
+                            argument.type().subtype().id() == ID_unsignedbv));
+      if(!arg_has_correct_type)
+      {
+        std::ostringstream error_message;
+        error_message << expr.source_location().as_string() << ": "
+                      << identifier << " has signature " << identifier
+                      << "(integral, integral, integral*), "
```

Comment:
GCC docs show: ` (type1 a, type2 b, type3 *res)` - implying a, b and res can all be *different* integral types, so might be worth writing this as `(integral1, integral2, integral3 *)` ? What does gcc give as an error message in this case?

Reply:
Version dependent I suppose. `gcc 10` says

```
test.c:4:46: error: argument 2 in call to function ‘__builtin_sub_overflow’ does not have integral type
    4 |   return __builtin_sub_overflow(1, (struct A){}, &x);
```

Reply:
Well, I prefer what you have than what GCC has :-) No big deal.